### PR TITLE
Allow to customize exporter and batching timers

### LINKED ISSUE #923: Geneva Exporter on Linux drops all logs from short-lived processes
# Issue with OpenTelemetry.Exporter.Geneva

List of [all OpenTelemetry NuGet packages](https://www.nuget.org/profiles/OpenTelemetry) and version that you are using (e.g. `OpenTelemetry 1.0.2`):

```
    <PackageReference Include="Microsoft.AspNetCore" Version="2.2.0" />
    <PackageReference Include="Microsoft.Extensions.Logging" Version="7.0.0" />
    <PackageReference Include="Microsoft.Extensions.Logging.Console" Version="7.0.0" />
    <PackageReference Include="OpenTelemetry.Exporter.Console" Version="1.3.2" />
    <PackageReference Include="OpenTelemetry.Exporter.Geneva" Version="1.3.1" />
```

It's happening with latest code built from tip of the `main` branch of contrib.

**Is this a feature request or a bug?**

It's a Feature Request to address a Bug..

* [X] Feature Request
* [X] Bug

**What is the expected behavior?**

I'm implementing a process restart monitor. No Logs should lost for short-lived processes, no matter how short-lived the process is - even if it lives for 1 second, all logs should be emitted and flushed to monitoring agent. You can compare the current behavior with Console Exporter that uses `SimpleLogRecordExportProcessor`, and that one doesn't drop any logs.

**What do you expect to see?**

All logs emitted by the process get passed to Geneva Monitoring Agent on Linux.

**What is the actual behavior?**

Exporter `Export` function is not getting called. No logs sent to Geneva. The issue is specifically affecting Unix Domain Socket, because Windows implementation uses different batching processor.

## Additional Context

I'd like to propose a patch to allow customizations of the batching parameters, by adding an additional delegate for creation of a batching processor with custom parameters:
```
    public static OpenTelemetryLoggerOptions AddGenevaLogExporter(
        this OpenTelemetryLoggerOptions options,
        Action<GenevaExporterOptions> configure,
        Func<BaseExporter<LogRecord>, BaseProcessor<LogRecord>> newProcessorDelegate = null)
    {
        Guard.ThrowIfNull(options);

        var genevaOptions = new GenevaExporterOptions();
        configure?.Invoke(genevaOptions);
        var exporter = new GenevaLogExporter(genevaOptions);

        if (newProcessorDelegate != null)
        {
            return options.AddProcessor(newProcessorDelegate(exporter));
        }

        if (exporter.IsUsingUnixDomainSocket)
        {
            return options.AddProcessor(new BatchLogRecordExportProcessor(exporter));
        }
        else
        {
            return options.AddProcessor(new ReentrantExportProcessor<LogRecord>(exporter));
        }
    }
```

That way the batching parameters can be configured irrespective of what OS the code is running on:

```
    builder.AddOpenTelemetry(options =>
    {
        options.AddGenevaLogExporter(exporterOptions =>
        {
             exporterOptions.ConnectionString = OpenTelemetryGenevaEndpoint;
             exporterOptions.CustomFields = new List<string> { "traceId", "eventName", "eventId", "componentId", "roleName" };
        },
        // Configure custom export processor: SimpleLogRecordExportProcessor or BatchLogRecordExportProcessor
        (exporter) => new BatchLogRecordExportProcessor(exporter,
                maxQueueSize : 2048,
                scheduledDelayMilliseconds : 100,
                exporterTimeoutMilliseconds : 1000,
                maxExportBatchSize : 1)
        ).
        SetResourceBuilder(ResourceBuilder.CreateDefault().AddService(serviceName: "MyService", serviceVersion: "1.0.0"));
    });
```

### PR DESCRIPTION
**Pull request by maxgolov**
Fixes #923

## Changes

Allow to customize batching processor, and/or batching processor timers.

**Usage**

New optional parameter added to `AddGenevaLogExporter` method. This parameter allows to supply a delegate to create a custom export processor. That way all logs emitted by process could show up at Unix Domain Socket sink in near-realtime rather than waiting for the next batching interval. This would be functionally similar to ETW behavior on Windows.

If new optional parameter is not supplied, then previous behavior with default instance of `BatchLogRecordExportProcessor` would take effect.

```
builder.AddOpenTelemetry(options =>
{
    options.AddGenevaLogExporter(exporterOptions =>
    {
        exporterOptions.ConnectionString = OpenTelemetryGenevaEndpoint;
        exporterOptions.CustomFields = new List<string> { "traceId", "eventName", "eventId", "componentId", "roleName" };
    },                            
    // Supply custom export processor: SimpleLogRecordExportProcessor or BatchLogRecordExportProcessor.
    // Below we configure near-realtime BatchLogRecordExportProcessor with delay of 100ms :
    (exporter) => new BatchLogRecordExportProcessor(exporter,
        maxQueueSize : 2048,
        scheduledDelayMilliseconds : 100,
        exporterTimeoutMilliseconds : 1000,
        maxExportBatchSize : 1)
    )
```

I am not sure if I need to modify the `PublicAPI.Shipped.txt`. This functionality could also be added by adding a completely new 3-parameters static function, i.e. not breaking the existing 2-parameters method.

### CODE REVIEW

File: src/OpenTelemetry.Exporter.Geneva/GenevaLoggingExtensions.cs [hunk 1]
```diff
@@ -24,13 +24,22 @@ namespace Microsoft.Extensions.Logging;
 
 public static class GenevaLoggingExtensions
 {
-    public static OpenTelemetryLoggerOptions AddGenevaLogExporter(this OpenTelemetryLoggerOptions options, Action<GenevaExporterOptions> configure)
+    public static OpenTelemetryLoggerOptions AddGenevaLogExporter(
+        this OpenTelemetryLoggerOptions options,
+        Action<GenevaExporterOptions> configure,
+        Func<BaseExporter<LogRecord>, BaseProcessor<LogRecord>> newProcessorDelegate = null)
```
Comment (cijothomas) [inline src/OpenTelemetry.Exporter.Geneva/GenevaLoggingExtensions.cs:30]: Which processor is used by the Exporter is an internal detail for this Exporter, and we don't want to expose more options for users to play with. The exporter relies on the specific type of processor being in the pipeline, and we don't expect end users to tweak it.

**Review by cijothomas — CHANGES_REQUESTED** [review]
See : https://github.com/open-telemetry/opentelemetry-dotnet-contrib/pull/924/files#r1085964172

Would like to keep config options minimum for this exporter, unless there are strong reasons.

File: src/OpenTelemetry.Exporter.Geneva/GenevaLoggingExtensions.cs (hunk 1 above)
Comment (maxgolov) [reply src/OpenTelemetry.Exporter.Geneva/GenevaLoggingExtensions.cs:30]: @cijothomas - please make it work reliably. Presently event loss is blocking us. This is a strong reason. I believe that Processors are intentionally exposed to devs working on custom exporters. If the fix to expose a delegate isn't the best fix, then could we please at least have the settings to configure the **batching** and the **flush** timers?

**Comment by reyang** [comment]
> That way all logs emitted by process could show up at Unix Domain Socket sink in near-realtime rather than waiting for the next batching interval. This would be functionally similar to ETW behavior on Windows.

@maxgolov would you elaborate why this is desired?

**Comment by CodeBlanch** [comment]
I think the underlying issue may be the socket doesn't flush its buffered data when disposed. See: https://github.com/open-telemetry/opentelemetry-dotnet-contrib/issues/923#issuecomment-1402680023

**Comment by cijothomas** [comment]
See this : https://github.com/open-telemetry/opentelemetry-dotnet-contrib/pull/925 It wont help with LogExporter yet, but it'll be added in 1.5 by the SDK itself, and can be extended to Geneva exporter too.

File: src/OpenTelemetry.Exporter.Geneva/GenevaLoggingExtensions.cs (hunk 1 above)
Comment (cijothomas) [reply src/OpenTelemetry.Exporter.Geneva/GenevaLoggingExtensions.cs:30]: See https://github.com/open-telemetry/opentelemetry-dotnet-contrib/pull/925 for ability to configure batch processor options. Only added for Traces now, will come for Logs in the 1.5.0 version.

Comment (maxgolov) [reply src/OpenTelemetry.Exporter.Geneva/GenevaLoggingExtensions.cs:30]: When can we expect something like this added to Geneva Logs exporter?

**Comment by maxgolov** [comment]
> @maxgolov would you elaborate why this is desired?

_Why realtime-ness for Event Logs is nice-to-have..._

@reyang - if you are debugging C# service code, you already instrument something with the `ILogger`. And if you use Console Exporter - you see the logs printed right away as you debug. Similarly, if you use something like `printf` - it lands in Docker container logs right away, shipped via [fluentd log driver right away by default](https://docs.docker.com/config/containers/logging/fluentd/)... The "async" buffered/batched mode is also available in Docker fluent logger driver, but **it is not the default**.

Similarly to that, developers debugging on Windows with ETW - get their events emitted right away. Those are even showing in Visual Studio Event Viewer right there, without any buffering.. Basically I'd like to see an option to control the batching on Linux. Since Unix Domain Socket is local, fast, with minimal connection time - what are the strong arguments for the logs batching, esp. with batching timer value so high that the logs are getting dropped? It seems like if batching is unavoidable, my scenario would work best with shorter batching timer. I'm not saying one hat fits all, thus it needs to be configurable.

**Comment by reyang** [comment]
> > @maxgolov would you elaborate why this is desired?
> 
> _Why realtime-ness for Event Logs is nice-to-have..._
> 
> @reyang - if you are debugging C# service code, you already instrument something with the `ILogger`. And if you use Console Exporter - you see the logs printed right away as you debug. Similarly, if you use something like `printf` - it lands in Docker container logs right away, shipped via [fluentd log driver right away by default](https://docs.docker.com/config/containers/logging/fluentd/)... The "async" buffered/batched mode is also available in Docker fluent logger driver, but **it is not the default**.
> 
> Similarly to that, developers debugging on Windows with ETW - get their events emitted right away. Those are even showing in Visual Studio Event Viewer right there, without any buffering.. Basically I'd like to see an option to control the batching on Linux. Since Unix Domain Socket is local, fast, with minimal connection time - what are the strong arguments for the logs batching, esp. with batching timer value so high that the logs are getting dropped? It seems like if batching is unavoidable, my scenario would work best with shorter batching timer. I'm not saying one hat fits all, thus it needs to be configurable.

@maxgolov thanks for the background. I think eventually GenevaExporter would move to use [user_events](https://docs.kernel.org/trace/user_events.html) on Linux, which is using kernel buffer, has blocking API exactly like what ETW has on Windows. So I was trying to understand how urgent/important this is, and do we consider the proposal/change as a solution or workaround or mitigation (e.g. changing the exporting frequency might reduce data loss but not prevent it). And if it is important but not urgent, I would suggest aligning the effort with user_events (e.g. one can imagine that if we moved to user_events, we won't use any batch exporting processor in GenevaExporter case).

**Comment by maxgolov** [comment]
@reyang - I have local patch, so this is not urgent. I just want to get a long-term fix to remove my patch :)

File: src/OpenTelemetry.Exporter.Geneva/GenevaLoggingExtensions.cs (hunk 1 above)
Comment (cijothomas) [reply src/OpenTelemetry.Exporter.Geneva/GenevaLoggingExtensions.cs:30]: after 1.4 release only.

**Comment by cijothomas** [comment]
closing as we are not proceeding with this.

**Pull request closed without merging**